### PR TITLE
HTTP/2 pipeline NIOAsyncChannel pipeline config

### PR DESCRIPTION
**Pull request by rnro**
Motivation:

This continues the work to expose functionality which allows users to interact with HTTP/2 connections via async abstractions and using structured concurrency.

We build atop of previous work to configure pipelines to deal with HTTP/2 (with no protocol negotiation) and to wrap connection/stream channels with `NIOAsyncChannel`s. This will allow users to iterate over streams and HTTP2Frames.

Modifications:

Provide functions which configure channels and pipelines with the HTTP2 handler, expose a multiplexer for dealing with streams and wrap connection and stream channels with `NIOAsyncChannel`s.

Result:

Users will be able to create and interact with HTTP/2 connections via `NIOAsyncChannel`s. Because HTTP/2 is a negotiated protocol and we do not yet handle it, this is of limited utility.

### CODE REVIEW

File: Sources/NIOHTTP2/HTTP2PipelineHelpers.swift [hunk 1]
```diff
@@ -418,6 +487,76 @@ extension Channel {
...
+            inboundStreamInitializer: inboundStreamInitializer
+        ).flatMap { multiplexer in
+            return connectionInitializer(self).flatMapThrowing { _ in
+                return try NIOAsyncChannel(
+                    synchronouslyWrapping: self,
+                    inboundType: ConnectionInbound.self,
+                    outboundType: ConnectionOutbound.self
+                )
+            }.map { connectionAsyncChannel in
+                return (connectionAsyncChannel, multiplexer)
```
Comment (glbrntt) [inline Sources/NIOHTTP2/HTTP2PipelineHelpers.swift:556]: we can get rid of the `map`:

```suggestion
                let connectionChannel = try NIOAsyncChannel(
                    synchronouslyWrapping: self,
                    inboundType: ConnectionInbound.self,
                    outboundType: ConnectionOutbound.self
                )

                return (connectionChannel, multiplexer)
```

File: Sources/NIOHTTP2/HTTP2PipelineHelpers.swift [hunk 2]
```diff
@@ -504,4 +643,60 @@ extension ChannelPipeline.SynchronousOperations {
...
+        mode: NIOHTTP2Handler.ParserMode,
+        connectionConfiguration: NIOHTTP2Handler.ConnectionConfiguration,
+        streamConfiguration: NIOHTTP2Handler.StreamConfiguration,
+        streamDelegate: NIOHTTP2StreamDelegate? = nil,
+        position: ChannelPipeline.Position = .last,
+        streamInboundType: StreamInbound.Type,
+        streamOutboundType: StreamOutbound.Type,
+        inboundStreamInitializer: @escaping NIOHTTP2Handler.StreamInitializer
+    ) throws -> NIOHTTP2Handler.AsyncStreamMultiplexer<NIOAsyncChannel<StreamInbound, StreamOutbound>> {
+        return try configureAsyncHTTP2Pipeline(
```
Comment (glbrntt) [inline Sources/NIOHTTP2/HTTP2PipelineHelpers.swift:686]: nit: explicit self

```suggestion
        return try self.configureAsyncHTTP2Pipeline(
```

File: Tests/NIOHTTP2Tests/ConfiguringPipelineAsyncMultiplexerTests.swift [hunk 3]
```diff
@@ -132,6 +132,91 @@ final class ConfiguringPipelineAsyncMultiplexerTests: XCTestCase {
...
+                connectionConfiguration: .init(),
+                streamConfiguration: .init(),
+                connectionInboundType: HTTP2Frame.self,
+                connectionOutboundType: HTTP2Frame.self,
+                streamInboundType: HTTP2Frame.FramePayload.self,
+                streamOutboundType: HTTP2Frame.FramePayload.self
+            ) { channel in
+                channel.eventLoop.makeSucceededVoidFuture()
+            } inboundStreamInitializer: { channel -> EventLoopFuture<Void> in
+                self.serverChannel.eventLoop.makeSucceededVoidFuture()
```
Comment (glbrntt) [inline Tests/NIOHTTP2Tests/ConfiguringPipelineAsyncMultiplexerTests.swift:151]: Don't mix up your event loops, this should be `channel.eventLoop`

File: Tests/NIOHTTP2Tests/ConfiguringPipelineAsyncMultiplexerTests.swift [hunk 4]
```diff
@@ -132,6 +132,91 @@ final class ConfiguringPipelineAsyncMultiplexerTests: XCTestCase {
...
```
Comment (glbrntt) [inline Tests/NIOHTTP2Tests/ConfiguringPipelineAsyncMultiplexerTests.swift:167]: Don't mix up your event loops, this should be `channel.eventLoop`

File: Tests/NIOHTTP2Tests/ConfiguringPipelineAsyncMultiplexerTests.swift [hunk 5]
```diff
@@ -132,6 +132,91 @@ final class ConfiguringPipelineAsyncMultiplexerTests: XCTestCase {
 
         serverRecorder.receivedFrames.assertFramePayloadsMatch(Array(repeating: ConfiguringPipelineAsyncMultiplexerTests.requestFramePayload, count: requestCount))
     }
+
+    func testNIOAsyncChannelPipelineCommunicates() async throws {
```
Comment (glbrntt) [inline Tests/NIOHTTP2Tests/ConfiguringPipelineAsyncMultiplexerTests.swift:136]: Can we have a test for the API which doesn't configure the connection channel as well?

File: Sources/NIOHTTP2/HTTP2ChannelHandler+InlineStreamMultiplexer.swift [hunk 6]
```diff
@@ -238,8 +238,36 @@ extension NIOHTTP2Handler {
...
         public func createStreamChannel<OutboundStreamOutput>(_ initializer: @escaping NIOHTTP2Handler.StreamInitializerWithOutput<OutboundStreamOutput>) async throws -> OutboundStreamOutput {
             return try await self.inlineStreamMultiplexer.createStreamChannel(initializer).get()
         }
+
+
+        /// Create a stream channel initialized with the provided closure and return it wrapped within a `NIOAsyncChannel`.
+        /// - parameters:
+        ///     - streamInboundType: The ``NIOAsyncChannel/inboundStream`` message type for outbound stream channels.
+        ///     This type must match the `InboundOut` type of the final handler added to the stream channel by the `initializer`
+        ///     or `HTTP2Frame.Payload` if there are none.
```
[the remark this answers is not in the document]
Comment (rnro) [reply Sources/NIOHTTP2/HTTP2ChannelHandler+InlineStreamMultiplexer.swift:251]: Changed to 
```
``HTTP2Frame/FramePayload``
```

File: Sources/NIOHTTP2/HTTP2ChannelHandler+InlineStreamMultiplexer.swift [hunk 7]
```diff
@@ -238,8 +238,36 @@ extension NIOHTTP2Handler {
...
+        ///     - streamOutboundType: The ``NIOAsyncChannel/outboundWriter`` message type for outbound stream channels.
+        ///     This type must match the `OutboundIn` type of the final handler added to the stream channel by the `initializer`
+        ///     or `HTTP2Frame.Payload` if there are none.
+        ///     - initializer: A callback that will be invoked to allow you to configure the
+        ///         `ChannelPipeline` for the newly created channel.
+        public func createStreamChannel<Inbound, Outbound>(
+            inboundType: Inbound.Type = Inbound.self,
+            outboundType: Outbound.Type = Outbound.self,
```
[the remark this answers is not in the document]
Comment (rnro) [reply Sources/NIOHTTP2/HTTP2ChannelHandler+InlineStreamMultiplexer.swift:259]: Yes, I agree, that was included by mistake 🙏

File: Sources/NIOHTTP2/HTTP2ChannelHandler+InlineStreamMultiplexer.swift [hunk 8]
```diff
@@ -238,8 +238,36 @@ extension NIOHTTP2Handler {
...
+        /// Create a stream channel initialized with the provided closure and return it wrapped within a `NIOAsyncChannel`.
+        /// - parameters:
+        ///     - streamInboundType: The ``NIOAsyncChannel/inboundStream`` message type for the created channel.
+        ///     This type must match the `InboundOut` type of the final handler added to the stream channel by the `initializer`
+        ///     or ``HTTP2Frame/FramePayload`` if there are none.
+        ///     - streamOutboundType: The ``NIOAsyncChannel/outboundWriter`` message type for the created channel.
+        ///     This type must match the `OutboundIn` type of the final handler added to the stream channel by the `initializer`
+        ///     or ``HTTP2Frame/FramePayload`` if there are none.
+        ///     - initializer: A callback that will be invoked to allow you to configure the
+        ///         `ChannelPipeline` for the newly created channel.
```
Comment (glbrntt) [inline Sources/NIOHTTP2/HTTP2ChannelHandler+InlineStreamMultiplexer.swift:256]: typo and formatting fixes:

```suggestion
        /// Create a stream channel initialized with the provided closure and return it wrapped within a `NIOAsyncChannel`.
        ///
        /// - Parameters:
        ///     - inboundType: The ``NIOAsyncChannel/inboundStream`` message type for the created channel.
        ///       This type must match the `InboundOut` type of the final handler added to the stream channel by the `initializer`
        ///       or ``HTTP2Frame/FramePayload`` if there are none.
        ///     - outboundType: The ``NIOAsyncChannel/outboundWriter`` message type for the created channel.
        ///       This type must match the `OutboundIn` type of the final handler added to the stream channel by the `initializer`
        ///       or ``HTTP2Frame/FramePayload`` if there are none.
        ///     - initializer: A callback that will be invoked to allow you to configure the
        ///         `ChannelPipeline` for the newly created channel.
```

**Review by glbrntt — COMMENTED** [review]
nit, looks good otherwise

**Review by glbrntt — APPROVED** [review]

File: Sources/NIOHTTP2/HTTP2ChannelHandler+InlineStreamMultiplexer.swift [hunk 9]
```diff
@@ -238,8 +238,37 @@ extension NIOHTTP2Handler {
...
+        /// - Parameters:
+        ///     - inboundType: The ``NIOAsyncChannel/inboundStream`` message type for the created channel.
+        ///       This type must match the `InboundOut` type of the final handler added to the stream channel by the `initializer`
+        ///       or ``HTTP2Frame/FramePayload`` if there are none.
+        ///     - outboundType: The ``NIOAsyncChannel/outboundWriter`` message type for the created channel.
+        ///       This type must match the `OutboundIn` type of the final handler added to the stream channel by the `initializer`
+        ///       or ``HTTP2Frame/FramePayload`` if there are none.
+        ///     - initializer: A callback that will be invoked to allow you to configure the
+        ///         `ChannelPipeline` for the newly created channel.
+        public func createStreamChannel<Inbound, Outbound>(
```
[the remark this answers is not in the document]
Comment (rnro) [reply Sources/NIOHTTP2/HTTP2ChannelHandler+InlineStreamMultiplexer.swift:258]: Resolved in https://github.com/apple/swift-nio-http2/pull/406

File: Sources/NIOHTTP2/HTTP2PipelineHelpers.swift [hunk 10]
```diff
@@ -258,6 +260,73 @@ extension Channel {
...
+    ///     - streamOutboundType: The ``NIOAsyncChannel/outboundWriter`` message type for inbound stream channels.
+    ///     This type must match the `OutboundIn` type of the final handler added to the stream channel by the `inboundStreamInitializer`
+    ///     or ``HTTP2Frame/FramePayload`` if there are none.
+    ///     - inboundStreamInitializer: A closure that will be called whenever the remote peer initiates a new stream.
+    /// - returns: An `EventLoopFuture` containing the `AsyncStreamMultiplexer` inserted into this pipeline which wraps
+    /// inbound streams as `NIOAsyncChannels` after initialization. The multiplexer can be used to initiate new streams
+    /// and iterate over inbound HTTP/2 stream channels.
+    @available(macOS 10.15, iOS 13.0, watchOS 6.0, tvOS 13.0, *)
+    @_spi(AsyncChannel)
+    public func configureAsyncHTTP2Pipeline<StreamInbound, StreamOutbound>(
```
[the remark this answers is not in the document]
Comment (rnro) [reply Sources/NIOHTTP2/HTTP2PipelineHelpers.swift:291]: Resolved in https://github.com/apple/swift-nio-http2/pull/406

**Pull request merged**